README.md
# regfile\_generics

[![BSD License][bsdlicense-button]][bsdlicense]
[![PyPI][pypi-button]][pypi]
[![Build Status][build-button]][build]
[![Coverage Status][codecov-button]][codecov]
[![Code style: black][black-button]][black]

[bsdlicense-button]: https://img.shields.io/github/license/icglue/regfile_generics
[bsdlicense]: https://opensource.org/license/bsd-2-clause/
[pypi-button]: https://img.shields.io/pypi/v/regfile_generics
[pypi]: https://pypi.org/project/regfile-generics/
[build-button]: https://github.com/icglue/regfile_generics/workflows/CI/badge.svg?event=push
[build]: https://github.com/icglue/regfile_generics/actions?query=workflow%3ACI+event%3Apush
[codecov-button]: https://codecov.io/gh/icglue/regfile_generics/branch/master/graph/badge.svg
[codecov]: https://codecov.io/gh/icglue/regfile_generics/tree/master
[black-button]: https://img.shields.io/badge/code%20style-black-000000.svg
[black]: https://github.com/psf/black

## Installation

### From PyPI

```bash
python3 -m pip install --upgrade regfile-generics
```

### From source

```bash
git clone https://github.com/icglue/regfile_generics
cd regfile_generics
python3 -m pip install .
```

## Usage

## Setup Regfile and Regfile Device for Access

See [tests/fixtures.py](https://github.com/icglue/regfile_generics/blob/master/tests/fixtures.py) how to create a regfile by deriving from the Regfile class.
Implement read/write functions and pass them (or overwrite while deriving) to an adequate [RegfileDevice](https://github.com/icglue/regfile_generics/blob/master/src/regfile_generics/regfile_device.py) class.

## Accessing Registers

```python
# dict like:
regfile["reg1_high"] = {"cfg": 0x0AA, "cfg_trigger": 0x0, "cfg_trigger_mode": 0x0}
# or single field (might issue read-modify-write)
regfile["reg1_high"]["cfg"] = 0xB

# uvm like (register have a _r suffix, field a _f suffix to avoid collisions):
regfile.reg1_high_r.cfg_f.set(2)
regfile.reg1_high_r.update()

# write_update
regfile["reg1_high"].write_update(cfg=0xA, cfg_trigger_mode=1)

# read (can be int or dict or string context)
print(regfile["reg1_high"])
assert (
    str(regfile["reg1_high"])
    == "Register reg1_high: {'cfg': 0xa, 'cfg_trigger': 0x0, 'cfg_trigger_mode': 0x1} = 0x1000a"
)
assert regfile["reg1_high"] == 0x1000A
assert dict(regfile["reg1_high"]) == {"cfg": 10, "cfg_trigger": 0, "cfg_trigger_mode": 1}

# read entire entry to a variable, so that no further read/write request will be issued
rh1 = regfile["reg1_high"].read_entry()
print(f"cfg: {rh1['cfg']}")
print(f"trigger: {rh1['cfg_trigger']}")
print(f"mode: {rh1['cfg_trigger_mode']}")

# bool access
regfile["reg1_high"] = 0
if regfile["reg1_high"]:
    assert False
```


CHANGELOG

v0.2.0
======

- add typing / drop python 3.8 support
- RegfileDev: deprecated readwrite_block
- Regfile: make `read/write` function private
- RegisterEntry: make `writable_field_items` function private
- add boolean conversion


tests/conftest.py
import pytest

from .fixtures import sessionmemregfile, sessionsimpleregfile, sessionsubwordregfile


tests/fixtures.py
#!/usr/bin/env python

import sys
import typing
from typing import cast

import pytest

from regfile_generics import Regfile, RegfileDevSimpleDebug, RegfileDevSubwordDebug, RegfileMemAccess

# pylint: disable=missing-class-docstring,missing-function-docstring


# Specific registerfile description
class SubmodRegfile(Regfile):
    def __init__(self, rfdev, base_addr):
        super().__init__(rfdev, base_addr)

        with self as regfile:
            with regfile["reg0"].represent(addr=0x0000, write_mask=0x0000001F) as reg:
                reg["cfg"].represent(bits="4:0", access="RW", reset="0x0", desc="Configure component")
                reg["status"].represent(bits="31:16", access="regfile", desc="Component status")

            with regfile["reg1_low"].represent(addr=0x0004, write_mask=0xFFFFFFFF) as reg:
                reg["cfg"].represent(bits="31:0", access="RW", reset="0x0", desc="Configure submod part 0")

            with regfile["reg1_high"].represent(addr=0x0008, write_mask=0x000301FF) as reg:
                reg["cfg"].represent(bits="7:0", access="RW", reset="0x0", desc="Configure submod part 1")
                reg["cfg_trigger"].represent(bits="8", access="RWT", reset="0b0", desc="trigger config update")
                reg["cfg_trigger_mode"].represent(
                    bits="17:16",
                    access="RWT",
                    reset="0b0",
                    desc="trigger config update",
                )

            with regfile["reg2"].represent(addr=0x00C0, write_mask=0x000001F0) as reg:
                reg["config"].represent(bits="8:4", access="RW", reset="0x0", desc="Configure component")
                reg["status"].represent(bits="31:16", access="regfile", desc="Component status")

            with regfile["reg_addr40"].represent(addr=0x0040, write_mask=0x0000001F) as reg:
                reg["start"].represent(bits="0", access="RW", reset="0b0", desc="start's the module")
                reg["enable_feature0"].represent(bits="1", access="RW", reset="0b1", desc="enables feature0")
                reg["enable_feature1"].represent(bits="2", access="RW", reset="0b0", desc="enables feature1")
                reg["enable_feature2"].represent(bits="3", access="RW", reset="0b1", desc="enables feature2")
                reg["enable_feature3"].represent(bits="4", access="RW", reset="0b0", desc="enables feature3")


# regfile devices could inherit from regfile_dev (-> implement rfdev_write // rfdev_read)
#     or its highlevel classes (which are simplifing implementation as:
#     - regfile_dev_simple  -> implementation of rfdev_write_simple(self, addr, value) // rfdev_read(self, addr)
#     - regfile_dev_subword -> implementation of rfdev_write_subword(self, addr, value, size) // rfdev_read(self, addr) [note that addr is unaligned(!) to byteper_word]


FixtureSubwordRegfile = typing.Annotated[tuple[Regfile, RegfileDevSubwordDebug], pytest.fixture]


@pytest.fixture(scope="session")
def sessionsubwordregfile() -> FixtureSubwordRegfile:
    regfile = SubmodRegfile(RegfileDevSubwordDebug(), 0xF000_0000)
    return cast(FixtureSubwordRegfile, (regfile, regfile.get_rfdev()))


FixtureSimpleRegfile = typing.Annotated[tuple[Regfile, RegfileDevSimpleDebug], pytest.fixture]


@pytest.fixture(scope="session")
def sessionsimpleregfile() -> FixtureSimpleRegfile:
    regfile = SubmodRegfile(RegfileDevSimpleDebug(), 0xF000_0000)
    return cast(FixtureSimpleRegfile, (regfile, regfile.get_rfdev()))


FixtureMemAccess = typing.Annotated[tuple[RegfileMemAccess, RegfileDevSimpleDebug], pytest.fixture]


@pytest.fixture(scope="session")
def sessionmemregfile() -> FixtureMemAccess:
    regfile = RegfileMemAccess(RegfileDevSubwordDebug(), 0xA000_0000, size=1024)
    return cast(FixtureMemAccess, (regfile, regfile.get_rfdev()))


tests/access_test.py
"""Regfile access tests"""

from pytest import warns

from .fixtures import FixtureMemAccess, FixtureSimpleRegfile, FixtureSubwordRegfile

# pylint: disable=line-too-long,missing-function-docstring


def test_dict_access(sessionsubwordregfile: FixtureSubwordRegfile) -> None:
    """dict like access"""
    regfile, rfdev = sessionsubwordregfile

    write_count = rfdev.write_count
    regfile["reg1_high"] = {"cfg": 0x0FF, "cfg_trigger": 0x1, "cfg_trigger_mode": 0x0}

    assert 0xF000_0008 in rfdev.mem.keys()
    assert rfdev.getvalue(0xF000_0008) == 0x1FF

    regfile["reg0"] = {"cfg": 0x11}

    read_count = rfdev.read_count
    reg0 = dict(regfile["reg0"])
    print(f"Debug: dict-read: {reg0}")

    assert reg0["cfg"] == 0x11
    assert read_count + 1 == rfdev.read_count
    assert rfdev.getvalue(0xF000_0000) & 0b11111 == 0x11

    assert write_count + 2 == rfdev.write_count


def test_str_cast(sessionsubwordregfile: FixtureSubwordRegfile) -> None:
    regfile, rfdev = sessionsubwordregfile

    write_count = rfdev.write_count
    read_count = rfdev.read_count
    regfile["reg1_high"] = {"cfg": 0x0AA, "cfg_trigger": 0x0, "cfg_trigger_mode": 0x0}

    print("Byte aligned - should no do read-modify-write...")
    regfile["reg1_high"]["cfg_trigger_mode"] = 0b11

    print("Reading register to string...")
    reg1_high_string = f"{regfile['reg1_high']}"
    print(reg1_high_string)
    expect_str = "Register reg1_high: {'cfg': 0xaa, 'cfg_trigger': 0x0, 'cfg_trigger_mode': 0x3} = 0x300aa"
    assert reg1_high_string == expect_str
    assert write_count + 2 == rfdev.write_count
    assert read_count + 1 == rfdev.read_count
    assert str(regfile["reg1_high"].field("cfg")) == "cfg"


def test_truncation_warning(sessionsubwordregfile: FixtureSubwordRegfile) -> None:
    regfile, rfdev = sessionsubwordregfile

    with warns(
        UserWarning,
        match=r"Field\(s\) cfg_trigger_mode were not explicitly set during write of register reg1_high!",
    ):
        regfile["reg1_high"] = {"cfg": 0x0BB, "cfg_trigger": 0x1}
    assert rfdev.getvalue(0xF000_0008) == 0x1BB

    with warns(UserWarning, match=r"^Ignoring non existent Field NOT_EXISTENT for write.$"):
        regfile["reg1_high"] = {"NOT_EXISTENT": 0x0FF, "cfg_trigger": 0x1}

    assert rfdev.getvalue(0xF000_0008) == 0x100

    with warns(
        UserWarning,
        match=r"^Writing read-only field status \(value: 0x00000100 -- mask: 0xffff0000 write_mask: 0x0000001f\).$",
    ):
        regfile["reg0"]["status"] = 0x100


def test_read_entry(sessionsubwordregfile: FixtureSubwordRegfile) -> None:
    regfile, rfdev = sessionsubwordregfile
    write_count = rfdev.write_count
    read_count = rfdev.read_count

    regfile["reg1_high"] = {"cfg": 0x011, "cfg_trigger": 0x0, "cfg_trigger_mode": 0b01}

    entry = regfile["reg1_high"].read_entry()
    assert rfdev.getvalue(0xF000_0008) == 0x10011

    entry["cfg"] = 0x22
    entry["cfg_trigger"] = 0x1
    entry["cfg_trigger_mode"] = 0x0
    assert entry["cfg"] == 0x22

    regfile["reg1_high"] = entry
    assert rfdev.getvalue(0xF000_0008) == 0x00122
    assert {"cfg": 0x077, "cfg_trigger": 0x0, "cfg_trigger_mode": 0b10} == dict(regfile["reg1_high"].get_dict(0x20077))

    assert regfile["reg1_high"].get_value({"cfg": 0x077, "cfg_trigger": 0x0, "cfg_trigger_mode": 0b10}) == 0x20077

    assert write_count + 2 == rfdev.write_count
    assert read_count + 1 == rfdev.read_count
    assert {"cfg": 0x22, "cfg_trigger": 0x1, "cfg_trigger_mode": 0x0} == dict(regfile["reg1_high"].get_dict())
    assert read_count + 2 == rfdev.read_count


def test_int_access(sessionsubwordregfile: FixtureSubwordRegfile) -> None:
    regfile, rfdev = sessionsubwordregfile

    read_count = rfdev.read_count
    write_count = rfdev.write_count

    regfile["reg1_low"] = 0xABCD1234
    assert regfile["reg1_low"]["cfg"] == 0xABCD1234

    regfile["reg1_high"] = 0x1234ABCD
    assert {"cfg": 0xCD, "cfg_trigger": 0x1, "cfg_trigger_mode": 0x0} == dict(regfile["reg1_high"])

    assert regfile["reg1_high"] == 0x1234ABCD
    assert int(regfile["reg1_low"]) == 0xABCD1234
    assert 0x1234ABCC < regfile["reg1_high"] < 0x1234ABCE
    assert 0x1234ABCD <= regfile["reg1_high"] <= 0x1234ABCE
    assert write_count + 2 == rfdev.write_count
    assert read_count + 8 == rfdev.read_count


def test_get_reset_values(sessionsubwordregfile: FixtureSubwordRegfile) -> None:
    regfile, rfdev = sessionsubwordregfile
    assert dict(regfile["reg_addr40"].get_reset_values()) == {
        "start": 0,
        **{f"enable_feature{i}": (i + 1) & 0x1 for i in range(4)},
    }


def test_rfdev_simple(sessionsimpleregfile: FixtureSimpleRegfile) -> None:
    regfile, rfdev = sessionsimpleregfile

    regfile["reg1_high"] = {"cfg": 0x07A, "cfg_trigger": 0x1, "cfg_trigger_mode": 0x0}

    write_count = rfdev.write_count
    read_count = rfdev.read_count

    assert rfdev.getvalue(0xF000_0008) == 0x0000_017A

    # read-modify-write
    regfile["reg1_high"]["cfg"] = 0x7B
    assert rfdev.getvalue(0xF000_0008) == 0x0000_017B

    # read-modify-write
    regfile["reg1_high"]["cfg_trigger_mode"] = 0x3
    assert rfdev.getvalue(0xF000_0008) == 0x0003_017B

    regfile["reg0"]["cfg"] = 0x1C

    assert rfdev.getvalue(0xF000_0000) & 0x1F == 0x1C

    assert read_count + 2 == rfdev.read_count
    assert write_count + 3 == rfdev.write_count


def test_mem(sessionmemregfile: FixtureMemAccess) -> None:
    regfile, rfdev = sessionmemregfile

    regfile.write_image(0x8, tuple(i for i in range(16)))

    for i in range(16):
        assert rfdev.mem[0xA000_0000 + 0x8 + i * rfdev.n_word_bytes] == i

    image = regfile.read_image(0xC, 15)
    for i in range(15):
        assert image[i] == i + 1


src/regfile_generics/regfile_device.py
# Copyright (C) 2020-2023 Andreas Dixius, Felix Neumärker
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

"""Generic Devices on which a regfile can operate."""

from __future__ import annotations

import logging
import os
import random
import struct
import sys
import traceback
import warnings
from typing import TYPE_CHECKING, cast

if TYPE_CHECKING:  # pragma: nocover
    from typing import Callable, Optional

    from .regfile import RegfileEntry


class RegfileDev:
    """Regfile Device class that handels the access of a Regfile

    :param callback: dict with rfdev_read/rfdev_write/blockread/blockwrite
                   pointing to a register read/write function
    :param bytes_per_word: bytes per word a single register access can handle (default 4)
    :param logger: logger instance
    :param prefix: prefix for (debug) logging with the logger instance

    .. deprecated:: 0.2.0

      :key blockread: reference to a blockread function, use callback dict instead
      :key blockwrite: reference to a blockwrite function, use callback dict instead"""

    def __init__(
        self,
        callback: Optional[dict[str, Callable]] = None,
        bytes_per_word: int = 4,
        logger: Optional[logging.Logger] = None,
        prefix: str = "",
        **kwargs,
    ):
        super().__init__()
        self._bytes_per_word = bytes_per_word
        self.logger = logger if logger else logging.getLogger(__name__)
        self._prefix = prefix
        self.callback = callback if callback else {}

        for blockop in ("blockread", "blockwrite"):
            if blockop in kwargs:  # pragma: nocover
                warnings.warn(
                    f"RegfileDev init - kwarg {blockop} has been deprecated use the callback dict instead.",
                    UserWarning,
                    stacklevel=2,
                )
                self.callback[blockop] = kwargs[blockop]

        if not isinstance(self.callback, dict):
            raise TypeError("Argument 'callback' has to dict with name, callback function.")

        if not set(self.callback) <= self._allowed_callbacks():
            raise AttributeError(f"Only {self._allowed_callbacks} are allowed as callback functions.")

        for func in self._allowed_callbacks() - set(self.callback):
            if not hasattr(self, func):
                raise TypeError(f"Function {func} has to be implemented or passed as callback function.")

    def _allowed_callbacks(self) -> set[str]:
        """Returns allowed keys to be passed as inside the callback argument while class initialization"""
        return {"rfdev_read", "rfdev_write"}

    @property
    def n_word_bytes(self) -> int:
        """Returs the number of bytes per word the device handles on one operation."""
        return self._bytes_per_word

    def blockread(self, start_addr: int, size: int) -> list[int]:
        """Initiate a blockread used for

        :param start_addr: start address of read data
        :param size: number of words to be read"""
        if "blockread" in self.callback:
            return self.callback["blockread"](start_addr, size)

        return [self.rfdev_read(start_addr + i * self.n_word_bytes) for i in range(size)]

    def blockwrite(self, start_addr: int, values: tuple[int, ...]) -> None:
        """Initiate a blockwrite used for memory access
        :param start_addr: start address of write data
        :param value: word list to be written"""

        if "blockwrite" in self.callback:
            self.callback["blockwrite"](start_addr, values)
            return

        mask = (1 << (8 * self.n_word_bytes)) - 1
        write_mask = mask
        for i, value in enumerate(values):
            self.rfdev_write(start_addr + i * self.n_word_bytes, value, mask, write_mask)

    def rfdev_read(self, addr: int) -> int:
        """Read method calling `rfdev_read` of callback dict passed upon init
        - could be overridden, when deriving a new RegfileDev"""
        return self.callback["rfdev_read"](addr)

    def read(self, baseaddr: int, entry: RegfileEntry) -> int:
        """Read a register entry relative to a base address

        :param baseaddr: Base address of the Register File
        :param entry: A register entry
        """
        value = self.rfdev_read(baseaddr + entry.addr)
        self.logger.debug(
            "%sRegfileDevice reading entry %s from address 0x%x = 0x%x",
            self._prefix,
            entry.name,
            entry.addr,
            value,
        )
        return value

    def rfdev_write(self, addr: int, value: int, mask: int, write_mask: int) -> None:
        """Read method could be overridden when deriving a new RegfileDev
        default implementation is to call the function passed while initialization,
        through the value of the "rfdev_write"-key of the callback dictionary.

        :param addr: Absolute Address to be accessed for write
        :param value: Value to be written
        :param mask: Mask for the write operation to the register
        :param write_mask: Mask of writeable bits inside the register
          (e.g. to determine if read-modify-write is necessary)
        """
        self.callback["rfdev_write"](addr, value, mask, write_mask)

    def write(self, baseaddr: int, entry: RegfileEntry, value: int, mask: int) -> None:
        """Read a register entry relative to a base address

        :param baseaddr: Base address of the Register File
        :param entry: A register entry
        :param value: register value
        :param mask:  mask for the operation
        """
        addr = baseaddr + entry.addr

        self.logger.debug(
            "%sRegfileDevice initiate write of entry %s at address "
            "0x%x -- {value: 0x%x, mask 0x%x, write_mask: 0x%x}",
            self._prefix,
            entry.name,
            entry.addr,
            value,
            mask,
            entry.write_mask,
        )
        self.rfdev_write(addr, value, mask, entry.write_mask)

    def readwrite_block(self, start_addr, values, write):  # pragma: nocover
        """.. deprecated:: 0.2.0

        Use :func:`blockread` or :func:`blockwrite` instead."""

        warnings.warn(
            "Function `readwrite_block()` is deprecated and will be removed in future versions.",
            UserWarning,
            stacklevel=2,
        )
        if not write:
            read_value = self.blockread(start_addr, len(values))
            for i in range(len(values)):  # pylint: disable=consider-using-enumerate
                # Modifications are be done by reference
                values[i] = read_value[i]
        else:
            self.blockwrite(start_addr, values)


class RegfileDevSimple(RegfileDev):
    """RegfileDev that operates on words only (implements read-modify-write if necessary)

    Derived from :class:`.RegfileDev`

    Allowed callback entries:

        :key rfdev_read: read function (signature :func:`.RegfileDev.rfdev_read`)
        :key rfdev_write_simple: write function which has the same signature like :func:`rfdev_write_simple`
    """

    def _allowed_callbacks(self) -> set[str]:
        return {"rfdev_read", "rfdev_write_simple"}

    def rfdev_write_simple(self, addr: int, value: int) -> None:
        """Simple write operation - calls back `rfdev_write_simple` if passed to constructor.

        :param addr: absolute address for write operation
        :param value: value to write
        """
        self.callback["rfdev_write_simple"](addr, value)

    def rfdev_write(self, addr: int, value: int, mask: int, write_mask: int) -> None:
        """:class:`.RegfileDev` rfdev_write implementations
        - executes the read-modify-write if necessary

        :param addr: absolute Address to be accessed for write
        :param value: value to be written
        :param mask: mask for the write operation to the register
        :param write_mask: mask of writeable bits inside the register"""

        keep_mask = ~mask & write_mask

        if keep_mask == 0:
            self.rfdev_write_simple(addr, value)
        else:
            # read, modify, write
            rmw_value = self.rfdev_read(addr)

            rmw_value &= ~mask
            rmw_value |= value & mask

            self.rfdev_write_simple(addr, rmw_value)


def regfile_dev_debug_getbits(interactive: bool, default_value: int, promptprefix: str) -> int:
    """Function to get bits for RegfileDebug* classes"""
    value = default_value
    if not interactive:
        print(f"{promptprefix} value: 0x{default_value:x}")
    else:  # pragma: nocover
        lasttrace: Optional[traceback.FrameSummary] = None
        regfile_generics_package_path = os.path.dirname(__file__)

        for stacktrace in traceback.extract_stack():
            if stacktrace[0].startswith(regfile_generics_package_path):
                if lasttrace is not None:
                    print(f"{lasttrace[0]}:{lasttrace[1]}: {lasttrace[3]}", file=sys.stderr)
                break
            lasttrace = stacktrace

        while True:
            if input_value := input(f"{promptprefix} value (0x{default_value:x}): "):
                try:
                    value = int(input_value, 0)
                    break
                except ValueError:
                    print(f"Invalid value {value}.")

    return value


class RegfileDevSimpleDebug(RegfileDevSimple):
    """Debug implementation of :class:`.RegfileDevSimple`

    :param interactive: if set to ``True`` the regfile device will request a user input upon read.
    """

    def __init__(self, interactive: bool = False, **kwargs):
        super().__init__(**kwargs)
        self.mem: dict[int, int] = {}
        self.write_count = 0
        self.read_count = 0
        self.__interactive = interactive

    def rfdev_read(self, addr: int) -> int:
        """Debug read function interactive if necessary

        :param: addr address
        """

        value = regfile_dev_debug_getbits(
            self.__interactive,
            self.getvalue(addr),
            f"{self._prefix}REGFILE-READING from addr 0x{addr:x}",
        )
        self.mem[addr] = value

        self.read_count += 1
        return value

    def rfdev_write_simple(self, addr: int, value: int) -> None:
        """Debug write function to `mem` attribute.

        :param addr: absolute address for write operation
        :param value: value to write
        """
        print(f"{self._prefix}REGFILE-WRITING to addr 0x{addr:x} value 0x{value:x}")
        self.mem[addr] = value
        self.write_count += 1

    def getvalue(self, addr: int) -> int:
        """Get value out of the memory, randomize if necessary.

        :param addr: address to be read"""
        if addr not in self.mem:
            value = random.getrandbits(8 * self.n_word_bytes)
            self.logger.info("Generating random regfile value 0x%x", value)
            return value

        return self.mem[addr]


class RegfileDevSubword(RegfileDev):
    """RegfileDev that operates on word sizes (implements read-modify-write if this)

    Derived from :class:`.RegfileDev`

    Allowed callback entries:

        :key rfdev_read: read function (signature :func:`.RegfileDev.rfdev_read`)
        :key rfdev_write_simple: write function which has the same signature like :func:`rfdev_write_simple`
    """

    def rfdev_write(self, addr: int, value: int, mask: int, write_mask: int) -> None:
        # register bits that must not be changed
        keep_mask = ~mask & write_mask

        # initial subword mask: full word size - all bits to 1
        subword_mask = (1 << (8 * self.n_word_bytes)) - 1

        # go from full word to shorter subwords (e.g. 32, 16, 8 bits --> 4, 2, 1 bytes)
        n_subword_bytes = self.n_word_bytes
        while n_subword_bytes:
            # iterate subwords of current size (e.g. 32bits: 0; 16bits: 0, 1; 8bits: 0, 1, 2, 3)
            for i in range(self.n_word_bytes // n_subword_bytes):
                # shift wordmask to current position
                i_word_mask = subword_mask << (i * n_subword_bytes * 8)

                # no keep bit would be overwritten? && all write bits are covered?
                if ((keep_mask & i_word_mask) == 0) and ((mask & (~i_word_mask)) == 0):
                    subword_offset = i * n_subword_bytes

                    # call virtual method
                    self.logger.debug(
                        "RegfileDevice: Subwrite address 0x%x -- {value: 0x%x, n_subword_bytes: 0x%x}",
                        addr + subword_offset,
                        value,
                        n_subword_bytes,
                    )
                    self.rfdev_write_subword(addr + subword_offset, value, n_subword_bytes)
                    # we are done here ...
                    return

            # reduce subword bytes - next half
            n_subword_bytes //= 2
            # reduce subword mask - throw away the other half
            subword_mask >>= n_subword_bytes * 8

        # no success?  - full read-modify-write
        rmw_value = self.rfdev_read(addr)

        rmw_value &= ~mask
        rmw_value |= value & mask

        # call virtual method
        self.rfdev_write_subword(addr, rmw_value, self.n_word_bytes)

    def _allowed_callbacks(self) -> set[str]:
        return {"rfdev_read", "rfdev_write_subword"}

    def rfdev_write_subword(self, addr: int, value: int, size: int) -> None:
        """Word size write operation - calls back `rfdev_write_subword` if passed to constructor.

        :param addr: absolute address for write operation (lower address bit indicate word position)
        :param value: value to write
        :param size: number of bytes to write
        """
        self.callback["rfdev_write_subword"](addr, value, size)


class RegfileDevSubwordDebug(RegfileDevSubword):
    """Debug implementation of :class:`.RegfileDevSimple`

    :param interactive: if set to ``True`` the regfile device will request a user input upon read.
    """

    def __init__(self, interactive: bool = False, **kwargs):
        super().__init__(**kwargs)
        self.__interactive = interactive
        self.mem: dict[int, int] = {}
        self.write_count = 0
        self.read_count = 0

    def getvalue(self, addr: int) -> int:
        """Return memory value

        :param addr: address were the data will be read
        """

        word = []
        for i in range(self.n_word_bytes):
            if (addr + i) not in self.mem:
                byte_value = random.getrandbits(8)
                print(f"Generating random regfile value {byte_value}.")
                self.mem[addr + i] = byte_value

            byte_value = self.mem[addr + i]
            word.append(byte_value)

        return int.from_bytes(struct.pack(f"{self.n_word_bytes}B", *word), "little")

    def rfdev_read(self, addr: int) -> int:
        """Debug read function interactive if necessary

        :param: addr address
        """

        value = self.getvalue(addr)

        value = regfile_dev_debug_getbits(
            self.__interactive,
            value,
            f"{self._prefix}REGFILE-READING from addr 0x{addr:x}",
        )
        for i in range(self.n_word_bytes):
            self.mem[addr + i] = (value >> (8 * i)) & 0xFF

        self.read_count += 1
        return value

    def rfdev_write_subword(self, addr: int, value: int, size: int) -> None:
        print(f"{self._prefix}REGFILE-WRITING to addr 0x{addr:x} value 0x{value:x} size=0x{size:x}")

        byte_values = value.to_bytes(self.n_word_bytes, "little")
        for i in range(size):
            self.mem[addr + i] = byte_values[i + addr & 0b11]
        self.write_count += 1


class StringCmdRegfileDevSimple(RegfileDevSimple):
    """Forwards regfile operations to a function call with string command,
    to do the regfile operations.

     Read:
         r<NUMBITS> <address>
             e.g. r32 0x1C

     Write:
         w<NUMBITS> <address> <value>
             e.g. w32 0x80 0xF9852A
    """

    def __init__(self, execute: Optional[Callable[[str], Optional[str]]] = None, **kwargs):
        super().__init__(**kwargs)
        self.execute = execute

    def rfdev_read(self, addr: int) -> int:
        """Debug read function implementation translates to :func:`execute()`

        :param addr: absolute address for read operation"""

        return int(cast(Callable[[str], str], self.execute)(f"r{8*self.n_word_bytes} 0x{addr:x}"), 0)

    def rfdev_write_simple(self, addr: int, value: int):
        """Debug write function implementation translates to :func:`execute()`

        :param addr: absolute address for write operation
        :param value: value to write"""

        cast(Callable[[str], None], self.execute)(f"w{8*self.n_word_bytes} 0x{addr:x} 0x{value:x}")


class StringCmdRegfileDevSubword(RegfileDevSubword):
    """Forwards regfile operations to a function call with string command,
    to do the regfile operations.

     Read:
         r<NUMBITS> <address>
             e.g. r32 0x1C

     Write:
         w<NUMBITS> <address> <value> [bsel]
             e.g. w32 0x80 0xF9852A
                  w32 0x80 0xF9852A 0x1
    """

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.execute = kwargs["execute"]

    def rfdev_read(self, addr: int) -> int:
        return int(self.execute(f"r{8*self.n_word_bytes} 0x{addr:x}"), 0)

    def rfdev_write_subword(self, addr: int, value: int, size: int) -> None:
        subword_addrbits = self.n_word_bytes.bit_length() - 1
        bsel = ((1 << size) - 1) << (addr & subword_addrbits)
        addr_aligned = addr & ~subword_addrbits

        self.execute(f"w{8*self.n_word_bytes} 0x{addr_aligned:x} 0x{value:x} 0x{bsel:x}")
